Disable indexing on OA fields that don't need to be searchable.

albertisfu opened the pull request:
In this PR I disabled indexing for some OA fields that don't need to be searchable, like `docket_slug`, `file_size_mp3`, etc. 

I checked the current Solr audio [schema](https://github.com/freelawproject/courtlistener-solr-server/blob/main/solr/cores/audio/conf/schema.xml) to confirm which fields shouldn't be indexed.

So this change will reduce the index size.

About stored fields, according to [documentation ](https://www.elastic.co/guide/en/elasticsearch/reference/current/mapping-store.html) fields are not stored by default and we don't need them to be stored since we get the values from the `_source` field as documentation suggests.

mlissner reviewed (commented):
Great! Do we need `court_exact` or `court_id` if we have `court.exact`? 

Also, what about `text`? We needed that for Solr, but do we need it for ES?

albertisfu commented:
@mlissner  about `court_id` and `court_exact`: We use `court_id` for filtering, so `court_exact` is not a mandatory field to be indexed. We could remove it from the document. However, it is part of the current API fields for OA search, so removing it might be problematic for some users. Maybe we should consider just disabling indexing for `court_exact`?
![Screenshot 2023-08-21 at 10 26 42](https://github.com/freelawproject/courtlistener/assets/486004/40dea232-c0f7-4aba-8458-221285d7b3ca)

Regarding `text`, we need it for the result snippet in the API and frontend. It is indexed since we're currently using it for searching within some fields that are not in the document mapping or suitable for text searching.

These are the fields that are available in the `text` field:

```
{# Audio file #}
{% if item.case_name_full %}
    {{ item.case_name_full }}
{% elif item.case_name %}
    {{ item.case_name }}
{% else %}
    {{ item.case_name_short }}
{% endif %}

{# Docket #}
{# Need dates so queries for the date are sure to be returned (see #271) #}
{{ item.docket.date_argued|date:"j F Y" }}
{{ item.docket.date_reargued|date:"j F Y" }}
{{ item.docket.date_reargument_denied|date:"j F Y" }}
{{ item.docket.docket_number }}

{# Transcript #}
{% if item.stt_status == 1 %}
    {{ item.transcript }}
{% endif %}

{# Court #}
{{ item.docket.court.full_name }}
{{ item.docket.court.citation_string }}
{{ item.docket.court.pk }}

{# Remainder of Audio File #}
{{ item.sha1 }}
{{ item.judges }}

```

**Case name:**
I submitted a fix in this PR to index the `caseName` field with the `best_case_name` (it was indexing only `case_name`), so now case name is not mandatory to be indexed in 'text'.

**Docket dates:**
We already have independent fields for the following dates in mapping.
```
date_argued
date_reargued
date_reargument_denied
```
However these fields are from the type `Date` so they can't be queried using a `query_string`.
So these fields are required to be indexed in the `text` field as strings. So that text query searches like:
`Johnson v. Potter Superior 1979` can return results, this is related to https://github.com/freelawproject/courtlistener/issues/271

**Docket number:**
We are currently searching within the `docket_number` as an independent field, so is not mandatory to be indexed in `text`.

**Transcript:**
I noticed we don't have transcripts for Audio documents yet. But if we expect to have them in the future, we could add an independent field called `transcript`, so it can be indexed and queried independently.

**Court:**
We're currently indexing and searching independently over
`court.full_name`
`court.citation_string`

The problem with `court.pk` is that it's of the type `Keyword`, so it can't be queried by a `query_string`. So if we want searches like `Johnson v. Potter Superior ca1` to return results, we need to continue indexing the `court.pk` as a string within text.

**sha1:**
Do we want users to be able to search for documents using the sha1?

**judges:**
We are currently indexing and searching over judges independently, so it's not required to be indexed within `text`.

**In summary:**
There are a few fields that are required to be indexed in `text` to be searchable using the search text box:
```
date_argued
date_reargued
date_reargument_denied
transcript
court_id
sha1 
```

We have some options here if we want to decrease the indexed size of `text`.

- We could create a new field, e.g., `indexed_text`, where we include only these fields. We can include this new field in the `query_string`, so these fields are searchable while disabling indexing from text and using it only for showing the snippet. 
  - The upside of this option is that we would only need to add one more field to the document mapping and one additional field to the `query_string`.
  - The downside is that we won't be able to control the weight of each of these fields in the search (or exclude one of the fields); we can set weight for the `indexed_text` field, but it will affect all the content within the field. This is how it currently works in Solr.

- Create independent text fields for: `date_argued`, `date_reargued`, `date_reargument_denied`, `transcript` and `court_id` so we can include them to the `query_string` making them searchable by the text box and disabling indexing from  `text`.
  - The upside of this option is that we'd be able to set an independent weight for each field for searching or excluding a field whenever we want without needing to re-index the `indexed_text` field.
  - The downside is that we need to add more fields to the document mapping, and the query string will become more complex since it will now need to look into more fields on each query.

- I also submitted a fix for keeping the `text` field in sync when other fields in the main document changed. 

Let me know what do you think, so I can apply the required changes.

mlissner commented:
> Maybe we should consider just disabling indexing for court_exact?

That seems right, yes.

> I submitted a fix in this PR to index the caseName field with the best_case_name (it was indexing only case_name), so now case name is not mandatory to be indexed in 'text'.

Hm, well, previously, it used `case_name_full` in `text`, but if we only use `best_case_name`, it'll often only use `case_name`. This means that searches for secondary parties (as opposed to lead parties) won't find hits that they would before. I.e., previously a search for "Saad" would pull up this document, but now it won't (because we only index the best case name, not the full one):

```
{
  case_name_full: Lissner, Saad, and A Boat v. Three Rubber Ducks, a Goose, and a Gander
  case_name: Lissner v. Three Rubber Ducks
}
```

> if we want searches like Johnson v. Potter Superior ca1 to return results, we need to continue indexing the court.pk as a string within text.

Yeah, that'd be good. I do that a lot, actually.

> Do we want users to be able to search for documents using the sha1?

It's useful once in a while, but I could live without it. Either way, I guess.

> Create independent text fields for....[a bunch of fields]

I think this is probably the better way to go. Something that bugs me about the current system is that snippets for OA are actually pretty useless. They don't have transcripts, so it's kind of a garbage bag of stuff we didn't put somewhere better. So...let's put it somewhere better. 

I think if we do that, we can also stop showing snippets in the OA search results unless we have transcripts. They're not useful at all. I mean, look at this:

![image](https://github.com/freelawproject/courtlistener/assets/236970/30b79d8b-0f9a-43bc-8bbe-3b740fecafa1)

That snippet is adding *nothing*.

When we get to doing RECAP and Opinions, we can extend this approach, using the text field to hold the text of the document, and (hopefully) putting the other things we're dumping into the `text` field into their own fields.

mlissner commented on cl/audio/tests.py:
```diff
@@ -66,7 +66,7 @@ def test_do_search_podcasts_have_content(self) -> None:
             reverse("search_podcast", args=["search"]),
             {
-                "q": f"court:{self.audio.docket.court}",
+                "q": f"court_id:{self.audio.docket.court.pk}",
```
This tweak surprised me a bit, but it looks like `court` wasn't a documented filter previously, so we're OK. One thing though, I noticed [in the documentation](https://www.courtlistener.com/help/search-operators/#fielded-queries-fieldname-term) that judges can still be filtered by `court` instead of `court_id`. Is that right and should we fix it? 

![image](https://github.com/freelawproject/courtlistener/assets/236970/5e02efe2-f0ac-4276-9cd6-f8b2a2af569b)

albertisfu replied:
Oh yeah, I changed this test because previously I removed all the fields in `audio_text.txt` and the tests regarding podcast in this PR were working on the Solr version. So this test was failing due to the court was not being indexed anymore, so I changed it to `court_id` which is the documented filter. 

I've updated this PR with it's base branch where Podcast were moved to ES. So this is not an issue anymore, however I still think the `court_id` filter is the right one in this test, right?

Then I restored content in `audio_text.txt` (which is now used only for Solr) since I realized will continue indexing OA documents in Solr for some time before we switch to ES completely, then we can remove the whole file.

About Judges, yeah we use `court_exact` for the courts filter, so maybe we should change it to `court_id` and adding it to documentation? `court` holds the court full name.

mlissner replied:
> I still think the court_id filter is the right one in this test, right?

Seems right to me.

> we should change it to court_id and adding it to documentation?

I agree.

mlissner reviewed (commented):
One comment for you. The rest looks good.

albertisfu commented:
@mlissner here are some comments and questions regarding the tweaks made in this PR.

> I think this is probably the better way to go. Something that bugs me about the current system is that snippets for OA are actually pretty useless. They don't have transcripts, so it's kind of a garbage bag of stuff we didn't put somewhere better. So...let's put it somewhere better.

Yeah, the content of `audio_text` is no longer indexed in the `AudioDocument`. Instead, some new fields were added to the document so that they are indexed and searchable. The new fields are:
`case_name_full, court_id_text, dateArgued_text, dateReargued_text, dateReargumentDenied_text`

I also made adjustments to the ES signal processor, ensuring that these text version fields are updated whenever the original fields are modified in the parent object.

`case_name_full`: I added this field to the document and included into the `query_string`, so it's searchable so the behaviour that you described is not an issue:
> Hm, well, previously, it used case_name_full in text, but if we only use best_case_name, it'll often only use case_name. This means that searches for secondary parties (as opposed to lead parties) won't find hits that they would before. I.e., previously a search for "Saad" would pull up this document, but now it won't (because we only index the best case name, not the full one)

`court_id_text`: The original field `court_id` is of type `Keyword` field so it's not searchable in the `query_string`, so this field version enables text queries that include the court_id.

`dateArgued_text,` dateReargued_text, dateReargumentDenied_text`: The original fields are `Date` fields, so they're not searchable using the query_string. So this text version holds the dates as text using the format it was using in audio_text `date:"j F Y"`, so searches by dates in text continue working.

`sha1`: This field was already being indexed; I simply changed it from a `Keyword` field to a `Text` field, making it now searchable in `query_strings`.

There were some issues when trying to stop indexing of `audio_text` while maintaining the ability to search across all the fields it held. To fix it some adjustments were made to the `query_string`.

The specific issue was as follows:

Let's consider the document.
```
case_name: SEC v. Frank J. Information
judge: John Smith
```

When multiple fields were indexed in the `text` field, it looked something like:
`text`: `... SEC v. Frank J. Information ... John Smith`

Since the` text` field was included in `query_string` fields alongside other fields.

A query like:  `SEC v. Frank Smith` returned the document as a result, since all the terms in the query were matched in the `text` field.

However when removing the content from the `text` and replaced with independent fields, the query look like this:

```
"should":[
            {
               "query_string":{
                  "fields":[
                     "court",
                     "court_citation_string",
                     "judge",
                     "caseName^5.0",
                     "docketNumber^2.0"
                  ],
                  "query":"SEC v. Frank Smith",
                  "quote_field_suffix":".exact",
                  "default_operator":"AND",
                  "tie_breaker":0.3
               }
            },
            {
               "query_string":{
                  "fields":[
                     "court",
                     "court_citation_string",
                     "judge",
                     "caseName^5.0",
                     "docketNumber^2.0"
                  ],
                  "query":"SEC v. Frank Smith",
                  "quote_field_suffix":".exact",
                  "default_operator":"AND",
                  "type":"phrase"
               }
            }
         ]
```
The query didn't bring back any results. The reason is that when you search across multiple fields using a `query_string`, the terms get expanded into an `OR` clause for each field. So, it becomes something like `field1:query_term OR field2:query_term`, as explained in the [docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-query-string-query.html#query-string-multi-field). This is why it didn't return the expected results.

So in this case the query look something like internally:
```
court:"SEC v. Frank Smith"
OR court_citation_string:"SEC v. Frank Smith"
OR judge:"SEC v. Frank Smith"
OR caseName:"SEC v. Frank Smith"
OR docketNumber:"SEC v. Frank Smith"
```
So the query doesn't return results since the search term `SEC v. Frank Smith` can't be found within any of the fields as it was in `text`, since the query it's composed by the `case_name` and `judge`.

So the workaround to this issue is also described in docs:

A query like this:
```
{
  "query": {
    "query_string": {
      "fields": [ "content", "name" ],
      "query": "this AND that"
    }
  }
}
```
Is internally interpreted as:

```
{
  "query": {
    "query_string": {
      "query": "(content:this OR name:this) AND (content:that OR name:that)"
    }
  }
}
```
So, each term in the query is searched in every field, looking for matches.

So, the solution involves adding an `AND` conjunction between each term in the query.
"SEC v. Frank Smith" -> "SEC AND v. AND Frank AND Smith"

However, an additional issue arose concerning scoring. Our text query is composed by two components of `query_string`: a `best_fields` query_string and a `phrase` query_string.

Introducing `AND` operators to both queries break the relevance ordering of search results since it impacted scoring.
The phrase `query_string` did not work as before as adding `AND` operators within each term break the formation of phrases.

For example these are the scores for the following query:
q:`Loretta E. Lynch`
Jose A. Dominguez v. Loretta E. Lynch  21.981869
Hong Liu Lorem v. Lynch-Loretta E.  10.313889
Hong Liu Yang v. Lynch-Loretta E.  10.313889

Now adding the ANDs these are the scores:
q: `Loretta AND E. AND Lynch`
Hong Liu Lorem v. Lynch-Loretta E.  20.058083
Hong Liu Yang v. Lynch-Loretta E.  20.058083
Jose A. Dominguez v. Loretta E. Lynch  20.058083

As you can seen the `phrase` query_string lost its relevancy and all the results get the same score.

So the solution to this issue was to only add the ANDs to the search terms in the `best_fields` query_string and keep the original terms for the `phrase` query_string:

q: `Loretta AND E. AND Lynch (best_fields)`
q: `Loretta E. Lynch (phrase)`
Jose A. Dominguez v. Loretta E. Lynch  21.981869
Hong Liu Lorem v. Lynch-Loretta E.  10.313889
Hong Liu Yang v. Lynch-Loretta E.  10.313889

So we get the same order and scores as before while the searching across multiple fields is solved.


**Unbalanced parenthesis bug:**

Working on the change above I noticed that a search like  `(Loretta AND Jose` or `(Loretta OR (SEC) AND Jose` or any other which unbalanced parenthesis break  the `query_string` leading to a Elasticsearch parsing error.

So I fixed by verifying all the parentheses in a query are balanced, in case there are unclosed parentheses the method removes the last unclosed one. 
 `(Loretta AND Jose` ->`Loretta AND Jose`
`(Loretta OR (SEC) AND Jose` -> `(Loretta OR SEC) AND Jose`

Do you think this approach is correct? I mean, removing the unclosed parentheses fixes the query, but it could also alter the original meaning of the query that the user initially had in mind. Would it be better to remove all the parentheses instead?

**Transcript:**
Now the transcript is indexed in the `text` field, it's also searchable and highlighted. It's only shown if it has content.

So now OA search results will look like thus:
![Screenshot 2023-08-24 at 10 21 07](https://github.com/freelawproject/courtlistener/assets/486004/780ad4e0-e74a-4b29-895a-c11044132499)

This change also affected the Alerts template:

Previously an alert for OA look like thus:
![Screenshot 2023-08-24 at 10 23 10](https://github.com/freelawproject/courtlistener/assets/486004/9b25c860-86b7-4a94-ad31-2481adec1b9e)

Now it look like thus:
![Screenshot 2023-08-24 at 10 23 45](https://github.com/freelawproject/courtlistener/assets/486004/5d006d2f-9cf7-46ea-8efe-58a3f3c26511)

If it has transcript I'll look like:
![Screenshot 2023-08-24 at 10 23 25](https://github.com/freelawproject/courtlistener/assets/486004/1d9e1f50-3521-437b-98b7-d7405b6b9a2b)

Does it looks correct? Do we need to add additional fields to the Alert?

**snippet in Search API:**
The snippet field in the search API for OA holds the content of the text field. Now that the text field contains the transcript. The snippet field in the API will be empty until transcripts are available. Do you think this can be a breaking change for some API users?

mlissner commented:
> So we get the same order and scores as before while the searching across multiple fields is solved.

Wow, this is complicated. Nice fix.

> I mean, removing the unclosed parentheses fixes the query, but it could also alter the original meaning of the query that the user initially had in mind. Would it be better to remove all the parentheses instead?

I think throwing a clear error is probably best:

> Unbalanced parentheses. Please correct the number of opening and closing parentheses in your query.

But it's also fine to fix them as you've done, provided the fixed query is returned to the user. The one thing we don't want to do is fix the query internally without somehow informing the user.

> Do [the search results and alerts] look correct? Do we need to add additional fields to the Alert?

I think so, yes. Those snippets weren't great.

> Do you think this can be a breaking change for some API users?

Yes, but it'd be weird. The field is really not very useful. I'll make a note of it in the release blog post I'm working on.

albertisfu commented:
> But it's also fine to fix them as you've done, provided the fixed query is returned to the user. The one thing we don't want to do is fix the query internally without somehow informing the user.

Got it, @ERosendo and I discussed on the best approach regarding this issue, and he suggested the following approach:

If we detect that the query is unbalanced, we won’t show any results and will explain to the user that the query is incorrect due to unbalanced parentheses. And we'll provide a suggestion (fixed parentheses) optional for the user to click and retrieve results for that query, or allow the user to fix the query on their own.

Eduardo, will be working on this approach while I continue reviewing comments and suggestions on the main OA PR.

mlissner commented:
> Eduardo, will be working on this approach while I continue reviewing comments and suggestions on the main OA PR.

Great. If providing the fixed response makes life harder, skip it! Gotta race to the finish line.

ERosendo commented:
@mlissner @albertisfu I've tweaked the search template to let users know the query is unbalanced and show the suggested query.

 Here are some screenshots:

<img width="1180" alt="Screenshot 2023-08-25 at 2 34 08 PM" src="https://github.com/freelawproject/courtlistener/assets/55959657/c036ce20-e425-4cc0-8539-dc1a7fcdd4c0">




<img width="1200" alt="Screenshot 2023-08-25 at 2 31 55 PM" src="https://github.com/freelawproject/courtlistener/assets/55959657/d90a9b28-74f2-4df7-9b21-d15055b34778">


Let me know what you think.

mlissner commented:
That looks great. Is there anything else in this one that we're waiting on or should I do another review and maybe merge?

albertisfu commented:
The other pending task here (beside the failing test that Eduardo is checking) was related to adding additional fields to OA Search Alerts in order to replace the snippet.

I have now added the Date Argued, Docket Number, Duration, and Judge fields to the template, and they now appear as follows:

HTML version:
![Screenshot 2023-08-25 at 17 25 02](https://github.com/freelawproject/courtlistener/assets/486004/b4e4f283-f130-4949-8f23-82d2beab1b5d)

Text version:
```
CourtListener.com
*****************

-------------------------------------------------------
We have news regarding your alerts at CourtListener.com
-------------------------------------------------------

Your real time oral argument alert -- Test Alert OA -- had 1 hit:
-------------------------------------------------------
Edit / Delete this Alert: https://www.courtlistener.com/alert/edit/490/
Disable this Alert (one click): https://www.courtlistener.com/alert/disable/RntoKkVWoCGRddvRDzYpd7B6OJQ6woyN8UU8NT2O/

1. RT Test OA (Bankr. C.D. Cal. 2023)
Date Argued: July 31st, 2023 | Docket Number: 19-5735 | Duration: 0 | Judge: John Smith
...RT Test OA transcript....
~~~~~
 - View this item on our site: https://www.courtlistener.com/audio/350/rt-test-oa/
 - Download original from the court: http://kirby.org/
 - Download the original from our backup: https://storage.courtlistener.com/dat/2023/08/01/example.dat


************************
This alert brought to you by the 501(c)(3) non-profit Free Law Project
 - Blog: https://free.law
 - Twitter: http://twitter.com/freelawproject
 - Donate: https://www.courtlistener.com/donate/?referrer=pt-email

Please donate to support our work.
```

So let me know what do you think about the template now.

mlissner commented:
Those look great, so I think it's just about fixing the broken test?

mlissner commented:
Merging, thank you both!
